alibi_detect/cd/keops/learned_kernel.py

```diff
@@ -0,0 +1,341 @@
+from copy import deepcopy
+from functools import partial
+from tqdm import tqdm
+import numpy as np
+from pykeops.torch import LazyTensor
+import torch
+import torch.nn as nn
+from torch.utils.data import DataLoader
+from typing import Callable, Dict, List, Optional, Union, Tuple
+from alibi_detect.cd.base import BaseLearnedKernelDrift
+from alibi_detect.utils.pytorch import get_device, predict_batch
+from alibi_detect.utils.pytorch.data import TorchDataset
+
+
+class LearnedKernelDriftKeops(BaseLearnedKernelDrift):
+    def __init__(
+            self,
+            x_ref: Union[np.ndarray, list],
+            kernel: Union[nn.Module, nn.Sequential],
+            p_val: float = .05,
+            x_ref_preprocessed: bool = False,
+            preprocess_at_init: bool = True,
+            update_x_ref: Optional[Dict[str, int]] = None,
+            preprocess_fn: Optional[Callable] = None,
+            n_permutations: int = 100,
+            batch_size_permutations: int = 1000000,
+            var_reg: float = 1e-5,
+            reg_loss_fn: Callable = (lambda kernel: 0),
+            train_size: Optional[float] = .75,
+            retrain_from_scratch: bool = True,
+            optimizer: torch.optim.Optimizer = torch.optim.Adam,  # type: ignore
+            learning_rate: float = 1e-3,
+            batch_size: int = 32,
+            batch_size_predict: int = 1000000,
```

Comment:
Tip: for numbers with at least 4 digits use underscores to visually be able to tell between e.g. a million and hundred thousand, i.e. `int = 1_000_000`.

Reply:
Assuming you've decided this is best left to do on a library-wide level later?